Userland/Libraries/LibPDF/Fonts/PS1FontProgram.cpp

```diff
@@ -0,0 +1,555 @@
+/*
+ * Copyright (c) 2022, Julian Offenhäuser <offenhaeuser@protonmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibPDF/CommonNames.h>
+#include <LibPDF/Encoding.h>
+#include <LibPDF/Fonts/PS1FontProgram.h>
+#include <LibPDF/Reader.h>
+#include <ctype.h>
+#include <math.h>
+
+namespace PDF {
+
+PDFErrorOr<void> PS1FontProgram::parse(ReadonlyBytes const& bytes, size_t cleartext_length, size_t encrypted_length)
+{
+    Reader reader(bytes);
+    if (reader.remaining() == 0)
+        return error("Empty font program");
+
+    reader.move_to(0);
+    if (reader.remaining() < 2 || !reader.matches("%!"))
+        return error("Not a font program");
+
+    if (!seek_name(reader, CommonNames::Encoding))
+        return error("Missing encoding array");
+
+    if (TRY(parse_word(reader)) == "StandardEncoding") {
+        m_encoding = Encoding::standard_encoding();
+    } else {
+        HashMap<u16, CharDescriptor> descriptors;
+
+        while (reader.remaining()) {
+            auto word = TRY(parse_word(reader));
+            if (word == "readonly") {
+                break;
+            } else if (word == "dup") {
+                u32 code_point = TRY(parse_int(reader));
+                auto name = TRY(parse_word(reader));
+                descriptors.set(code_point, { name.starts_with('/') ? name.substring_view(1) : name.view(), code_point });
+            }
+        }
+        m_encoding = TRY(Encoding::create(descriptors));
+    }
+
+    bool found_font_matrix = seek_name(reader, "FontMatrix");
+    if (found_font_matrix) {
+        auto array = TRY(parse_number_array(reader, 6));
+        m_font_matrix = { array[0], array[1], array[2], array[3], array[4], array[5] };
+    } else {
+        m_font_matrix = { 0.001f, 0.0f, 0.0f, 0.001f, 0.0f, 0.0f };
+    }
+
+    auto decrypted = TRY(decrypt(reader.bytes().slice(cleartext_length, encrypted_length), 55665, 4));
+    return parse_encrypted_portion(decrypted);
+}
+
+Gfx::Path PS1FontProgram::build_char(u32 code_point, Gfx::FloatPoint const& point, float width)
+{
+    if (!m_glpyh_map.contains(code_point))
+        return {};
+
+    auto glyph = m_glpyh_map.get(code_point).value();
+    auto scale = width / m_font_matrix.map(glyph.width).x();
+    auto transform = m_font_matrix;
+
+    // FIXME: Why do we need to flip the glyph here?
```

Comment:
I'm assuming this is because the Y coordinate of a PDF increases from the bottom to the top of the page, whereas in Serenity (and most graphics in general) it increases from top to bottom. `Renderer` [does something similar](https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibPDF/Renderer.cpp#L47).

Reply:
You're right! I forgot that we convert character space (that goes from bottom to top like _user space_) directly to device space, the pixel grid here. I updated the comment :^)